### DISCUSSION TITLE
Multi touch pinch in Avalonia on Windows

### DISCUSSION THREAD

**Discussion by mgnslndh** [AvaloniaUI/Avalonia#9651] [Q&A] [answered]
It now seems like multi touch is enabled by default if supported by Windows. There is also an API for gestures but I do not find any implementation for a pinch gesture.

- Is there any plans on adding such a gesture?
- Is there is any example code that implements any kind of pinch support in an Avalonia app?

I'm specifically looking to use touch pinch in combination with https://github.com/wieslawsoltes/PanAndZoom

**Comment by maxkatz6** [comment, marked as answer]
> Is there any plans on adding such a gesture?

No plans in the backlog of the core team, but it would be great to have an optional gesture recognizer for that.

> Is there is any example code that implements any kind of pinch support in an Avalonia app?

Not sure, if we have any. But it should be implemented using APIs available in the IGestureRecognizer, like we have here https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Base/Input/GestureRecognizers/ScrollGestureRecognizer.cs.

**Reply by mgnslndh** [reply]
OK, thanks for the feedback!

**Comment by mgnslndh** [comment]
This PR was recently added and will, among other things, add a PinchGestureRecognizer.
https://github.com/AvaloniaUI/Avalonia/pull/9751

I will check out the CI build and see how it works.

**Reply by mgnslndh** [reply]
The `PinchGestureRecognizer` works fine on its own but I have not yet got it to work with the PanAndZoom library. Might have to wait until a proper release of Avalonia.

**Reply by jhimes144** [reply]
@mgnslndh 

I haven't yet attempted this, but I plan on implementing touch zoom for the PanAndZoom library in my project later down the road from mobile. What exactly are the issues you are running into?

**Reply by mgnslndh** [reply]
I have tested the pinch gesture and it seems to work nicely. I then did a quick test to see if I could use that gesture with the latest version of PanAndZoom. I did not work because of nuget package version incompatibilities but I guess that cloning PanAndZoom and implementing the fix it should work.

**Reply by mgnslndh** [reply]
Touch zoom in PanAndZoom would be a great contribution to that project. I'd be happy to test it.

**Reply by jhimes144** [reply]
I thought of implementing my own, I will if their isn't a solution available once I get around to that part of my app, I'll likely contribute it to the PanZoom project. 

In the meantime it should be pretty easy to create your own handles off the ZoomPan control events and implement the math found in the zoom pinch gesture code you linked. I think it could be implemented fairly quickly, a short evening's work.
